x/data/server/lookup/table.go

```diff
@@ -0,0 +1,159 @@
+package lookup
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"hash/fnv"
+)
+
+// Table is the interface for interacting with a lookup table.
+type Table interface {
+	// GetOrCreateID is an idempotent method for creating or retrieving an unique
+	// shortened identifier for the provided binary value.
+	GetOrCreateID(store KVStore, value []byte) []byte
+
+	// GetValue returns the binary data (if any) corresponding to the provided shortened identifier.
+	GetValue(store KVStore, id []byte) []byte
+}
+
+// NewTable creates a new lookup table for the provided, optional KVStore prefix using default parameters.
+// Default parameters are currently set to use the first 4-bytes of the FNV-1a 64-bit, non-cryptographic hash.
+// In the case of a collision, more bytes of the hash will be used for disambiguation but this happens
+// in a minority of cases except for massively large data sets.
+func NewTable(prefix []byte) (Table, error) {
+	return NewTableWithOptions(TableOptions{
+		Prefix: prefix,
+	})
+}
+
+// NewTableWithOptions creates a Table with custom options. Most users should just use NewTable
+// with the default values.
+func NewTableWithOptions(options TableOptions) (Table, error) {
+	prefixLen := len(options.Prefix)
+	minLength := options.MinLength
+	if minLength == 0 {
+		minLength = 4
+	}
+
+	newHash := options.NewHash
+	if newHash == nil {
+		newHash = func() hash.Hash {
+			return fnv.New64a()
```

Comment:
How about just blake hashes which are in the go stdlib? Is there a concern in adding a dependency on SipHash? I guess not since it's in BTC? I'm fine to use whatever is the best, most peer-reviewed hash function for this purpose.

Reply:
`SipHash` is widely used for hash map implementation. It's a small dependency. It could create a standard push back (to use only stdlib hashes). BTW, `SipHash` is designed by  Jean-Philippe (he is a prof at EPFL - neighbor university), which is an author of `blake` family functions. I can ask him - was chatting with him in the past.

Reply:
Since this is a dependency on #236, how about we address this as a follow up PR after we've decided on the correct hash function and implementation?

I'm planning to address the other comments and then it would be good to merge so #236 can get unblocked.

Reply:
yes, let's just create a task.

Reply:
https://github.com/regen-network/regen-ledger/issues/248